### PR TITLE
External Media: Implementation

### PR DESCRIPTION
**Pull request by retrofox**
External Media allows users to browse and insert [free images from Pexels](https://wordpress.com/support/free-photo-library/) or [their own Google Photos library](https://wordpress.com/support/google-photos/).

This feature extends the Media Library button in block like Image, Gallery, etc, making it easy to find and convenient to use:

81190858-a4ceeb80-8fb8-11ea-992e-5339e9e73aac

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
It's a feature new to Jetpack but existing on Dotcom.
See 47-gh-dotcom-private

#### Testing instructions
* This PR adds a new npm module, please run `yarn && yarn run build-extensions` to pull it in.
* Open the editor and insert an image block.
* Click the Select Image button and choose either of the new libraries.
* For Pexels:
  * Change the search phrase to explore how the results update.
  * Click the Load more button and make sure the content looks right and gets inserted in the right place
  * Select an image and click "Copy & Insert"
  * Make sure the image is copied into your media library and inserted properly.
* For Google Photos:
  * Navigate to https://wordpress.com/marketing and select your test site.
  * Make sure Google Photos is disconnected.
  * Back in the Editor: Select Google Photos from the button dropdown
  * Make sure it's asking you to connect to your Google account
  * After successfully connecting, select a photo album or selectet Recent Photos from the dropdown.
  * Select an image and click "Copy & Insert"
  * Make sure the image is copied into your media library and inserted properly.

Also see demo video in paYJgx-G1-p2.

#### Proposed changelog entry for your changes:

* Adds support for Pexels and Google Photos media libraries.

### CODE REVIEW

**Comment by obenland** [comment]
To what extend have you tried repurposing the API work from the original block extension? It looks like they're using the existing v1.1 endpoint there.

Would it be easer to wait for #15721 to be in a working condition to be able to test these endpoint more easily? I've had a hard time getting this PR to work with a fully functioning and connected Jetpack site.

**Comment by retrofox** [comment]
> To what extend have you tried repurposing the API work from the original block extension? It looks like they're using the existing v1.1 endpoint there.

I'm using almost the same logic implemented in the v1.1 version (WP COM API). The issue is it isn't possible to get the user ID when the request is made from the Jetpack site, through the wpcom_json_api_request_as_user()  helper function. The easier way to do it is by adding a new point to WP REST API.

> Would it be easer to wait for #15721 to be in a working condition to be able to test these endpoint more easily? I've had a hard time getting this PR to work with a fully functioning and connected Jetpack site.

Yes, probably. It gets tricky because you need to be authenticated to test it. I created a request in the client in order to test it. Anyway, let me try to find out a way to test it easily.

I suggest using this one as the feature branch and try to branch off from here.

File: _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/list/' . self::$services_regex,
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_external_media' ),
+				'permission_callback' => array( $this, 'permission_callback' ),
+				'args' => array(
+					'search' => array(
+						'description' => __( 'Media collection search term' ),
```
Comment (jeherve) [inline _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:42]: You'll want to add the Jetpack textdomain here.

```suggestion
						'description' => __( 'Media collection search term', 'jetpack' ),
```

**Comment by marekhrabe** [comment]
We need these proxy endpoints return the same http status code as the original endpoint did. 

There is a logic established in the frontend code for error handling (for example the google photos requiring authorization), which is currently ignored because all responses are coming back as 200 - success.

File: _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php [hunk 2]
```diff
@@ -249,7 +249,7 @@ public function get_connection_details( \WP_REST_Request $request ) {
 		$service    = rawurlencode( $request->get_param( 'service' ) );
 		$wpcom_path = sprintf( '/meta/external-services/%s', $service );
 
-		$response = Client::wpcom_json_api_request_as_user( $wpcom_path, '2', array(), null, 'rest' );
+		$response = Client::wpcom_json_api_request_as_user( $wpcom_path, '1.1', array(), null, 'rest' );
```
Comment (retrofox) [inline _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:252]: 🤔 have we changed these values before?

Comment (obenland) [reply _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:252]: Possibly? In my testing just now `2` didn't give me a result, though, `1.1` did.

Comment (retrofox) [reply _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:252]: Oh, I see it was [modified here](https://github.com/Automattic/jetpack/commit/906f2d43e575ea2d8e16c925aeb58e451b9d29bd#diff-6c5528991a5cda0f1a07c07e5e9fd9f3R249).

**Comment by brbrr** [comment]
With this PR, some blocks (tiled gallery, simple payments) breaks on edit: `TypeError: Cannot call a class as a function`

The error comes from AMP plugin, so it might be AMP compatibility issue. Rebased, to make sure it's not stale `master` issue.

**Comment by marekhrabe** [comment]
@obenland shared this yesterday:

> After Fusion was fixed, I tested https://github.com/Automattic/jetpack/pull/15717 on dotcom (https://code.a8c.com/D43601). Unfortunately things didn't go smoothly. 
> https://github.com/Automattic/jetpack/pull/15717/commits/2f11800d817ebace9bcef886d312b6ed3816d7c7 fixes the feature for dotcom, but it seems like it broke api response handling for Jetpack.

I believe that's what broke the API. I'll try looking into it but I'm not very proficient in that area

> Also, I wonder if it worth doing a CFT for this feature.

I couldn't think of reason why we won't have one. We are always doing those

> 2300 additions are rather a lot.

Agreed - the code was "done" elsewhere and then ported here by us.

**Comment by marekhrabe** [comment]
AMP plugin compatibility should be ok now.

**Comment by marekhrabe** [comment]
I have identified 44ab6b6697cb5494776072d59a9b6bba701d3e58 as the breaking commit that prevents this PR to work in Jetpack. It contains some code in PHP which is guarded with `IS_WPCOM` checks but there are some extra changes in JS which are applied to all requests and it leads to errors. 

One example being `parse: false,`

https://github.com/Automattic/jetpack/pull/15717/commits/44ab6b6697cb5494776072d59a9b6bba701d3e58#diff-257ae69b41152e8f75d694704b9193d0R125

which means that response is no longer parsed as JSON (which the handler expects) but it gets passed as the response object from `apiFetch` with data still stringified in the `body` property of passed object.

**Comment by obenland** [comment]
Apologies, I should have posted the update in-ticket and updated the labels.

Testing this PR on dotcom with D43601-code I noticed some inconsistencies with the way `wpcomFetch` processes POST requests and error responses. The changes in https://github.com/Automattic/jetpack/commit/44ab6b6697cb5494776072d59a9b6bba701d3e58 fixed those, but I didn't get to correct the errors it introduced in Jetpack before the end of my day.

I'm still trying to get my local ngrok working again after the recent changes there. Until I have that working again, replacing `parse: false` with `parse: window.wpcomFetch === typeof undefined` will hopefully make it work for both environments

File: extensions/shared/external-media/constants.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		label: __( 'Weddings', 'jetpack' ),
+	},
+];
+export const PEXELS_EXAMPLE_QUERIES = [
+	'mountain',
+	'ocean',
+	'river',
+	'clouds',
+	'pattern',
+	'shapes',
```
Comment (obenland) [inline extensions/shared/external-media/constants.js:135]: @marekhrabe When you get a chance, would you mind removing this term? It gives me unexpected search results.

**Comment by marekhrabe** [comment]
Named functions in `memo()` calls are intentional so the React devtools and stack traces show the name and not just "Unknown" for all components. Bart has suggested a syntax update to change how we call memo and will implement it

**Comment by WunderBart** [comment]
> Named functions in `memo()` calls are intentional so the React devtools and stack traces show the name and not just "Unknown" for all components. Bart has suggested a syntax update to change how we call memo and will implement it

Updated in https://github.com/Automattic/jetpack/pull/15717/commits/93b9a498ec89d992fc7afd89e5d91acf35598ab6.

**Comment by obenland** [comment]
D43601-is update with the latest changes and built has gone through 🟢 .

@Automattic/ajax Please hold off from merging any more follow-up PRs into this feature branch until this PR has been approved and merged itself. We're four days away from code-freeze for Jetpack 8.6, let's give @jeherve & @brbrr a chance to review this without more changes.

**Comment by jancavan** [comment]
@retrofox The hierarchy of the buttons don't look right to me. I vaguely remember coming across a mockup by @jasmussen that had primary, secondary, and tertiary buttons. Was that not implementable?

**Comment by obenland** [comment]
> Was that not implementable?

Design-wise we took what was there when we ported the feature over from wpcom-blocks.

In the interest of shipping and iterating, let's not make any more changes here and rather address any follow-up design updates in separate issues and PRs.

**Comment by jancavan** [comment]
> In the interest of shipping and iterating, let's not make any more changes here and rather address any follow-up design updates in separate issues and PRs.

@obenland Yes, definitely. Sorry for the confusion. What I was referring to was that - the buttons in this PR (and in my recent test), there are 2 secondary buttons separated by a regular link:

Screen Shot 2020-05-20 at 3 16 06 PM

Just wanted to make sure it reflects the image below:

Screen Shot 2020-05-20 at 3 15 41 PM

**Comment by obenland** [comment]
I created #15874 to review the button design. Let's make any necessary changes in a follow up PR so this one can move forward in the review process.

File: extensions/shared/external-media/media-browser/media-item.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	}, [ props.onClick ] );
+
+	const { item, isSelected, isCopying = false } = props;
+	const { thumbnails, caption, name, title, type, children = 0 } = item;
+	const { medium = null, fmt_hd = null } = thumbnails;
+	const alt = title || caption || name;
+	const classes = classnames( {
+		'jetpack-external-media-browser__media__item': true,
+		'jetpack-external-media-browser__media__item__selected': isSelected,
```
[the remark this answers is not in the document]
Comment (jeherve) [reply extensions/shared/external-media/media-browser/media-item.js:34]: We try to follow the same guidelines as Calypso's, but you'll find that this is not always strictly respected:
https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md#general-syntax-and-writing-rules

File: extensions/shared/external-media/sources/google-photos/filter-view.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ */
+import { __ } from '@wordpress/i18n';
+import { Fragment, useState } from '@wordpress/element';
+import { SelectControl, Button } from '@wordpress/components';
+
+const FILTERS = [
+	{ label: __( 'Category', 'jetpack' ), value: 'category' },
+	{ label: __( 'After date', 'jetpack' ), value: 'startDate' },
+	{ label: __( 'Before date', 'jetpack' ), value: 'endDate' },
+	{ label: __( 'Favourite Media Only', 'jetpack' ), value: 'favorite' },
```
Comment (jeherve) [inline extensions/shared/external-media/sources/google-photos/filter-view.js:12]:
```suggestion
	{ label: __( 'Favorite Media Only', 'jetpack' ), value: 'favorite' },
```

Comment (WunderBart) [reply extensions/shared/external-media/sources/google-photos/filter-view.js:12]: I'd even suggest using the same wording as in the added filter - "Only favorites":
   Screenshot 2020-05-21 13 55 23

File: extensions/shared/external-media/media-browser/media-item.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+function MediaItem( props ) {
+	const onClick = useCallback( () => {
+		if ( props.onClick ) {
+			props.onClick( props.item );
+		}
+	}, [ props.onClick ] );
+
+	// Catch space and enter keypresses.
+	const onKeyDown = event => {
+		if ( event.which === 13 || event.which === 32 ) {
```
Comment (WunderBart) [inline extensions/shared/external-media/media-browser/media-item.js:21]: `which` is deprecated and might not be supported everywhere. We should use `key` instead: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which

File: extensions/shared/external-media/media-browser/media-item.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (WunderBart) [reply extensions/shared/external-media/media-browser/media-item.js:34]: Thanks for the source. Looks like we should be using the `is-*` convention:

> This is an example of a declaration:
> 
> `.component__element.is-modifier {}`
>
> (...)
> - The .is-modifier class should never be written on its own outside of the component class. This avoids naming collisions and provides a consistent use of modifiers.

Will keep that in mind!

**Comment by jancavan** [comment]
One small thing - we are ok using the Pexels logo as an icon, right? I could've sworn I stumbled upon an old thread saying we couldn't use it, hence we ended up with a generic icon for it in Calypso. But maybe I'm not remembering it correctly. @jasmussen Do you know?

**Comment by jasmussen** [comment]
I believe that was for Google Photos, but that this changed to allow it. So I believe we should be good!

File: extensions/shared/external-media/media-browser/media-item.js (hunk 6 above)
Comment (obenland) [reply extensions/shared/external-media/media-browser/media-item.js]: Fixed in 95f6109

**Comment by obenland** [comment]
Thank you for your feedback @jeherve, the issues you pointed out should be fixed now!

> When I first use the options, I would expect to go through a connection flow of some sort:

I wonder if that was because authentication handler package wasn't pulled in? I updated the testing instructions to ask for `yarn` to be run first, so [that new npm module](https://github.com/Automattic/jetpack/pull/15717/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R93) gets pulled in.

> Should Pexels work out of the box with no connection?

It should work out of the box when the site is connected to WP.com. There shouldn't be any additional authentication necessary.

**Comment by jancavan** [comment]
Tiny nitpick - I think it should be "Pexels free photos". We've started to standardize the use of sentence-case in many areas:

https://github.com/Automattic/wp-calypso/pull/39692
https://github.com/Automattic/wp-calypso/pull/41647

File: extensions/shared/external-media/media-browser/media-item.js (hunk 6 above)
Comment (WunderBart) [reply extensions/shared/external-media/media-browser/media-item.js]: See https://github.com/Automattic/jetpack/commit/95f6109c6b3ce8800143494c1db4f4d7611978d0#r39443020.

**Comment by jasmussen** [comment]
> In a V2, I think it would be nice to make that "Load More" button actually load more images as you scroll down:

This needs to be considered in light of keyboard accessibility. We actually had infinite scroll and I added that button so that keyboard users can actually tab to the "insert" button, as opposed to tabbing endlessly through an ever-expanding list. An alternate approach might work, but agree it seems like a followup.

> The media insert modal always has scrollbars, which seems a bit weird:

I'd love to hepl here if I can. One good method for debugging this is to open the console in Chrome, and type the following:

```
setTimeout(function(){debugger;}, 2000);
```

Once you hit Enter, you have 2 seconds to invoke the "copying" dialog in question, at which point Chrome will lock the debugger. Getting the timing right can be a little tricky, but once you get it, you can comfortably use the inspector to figure out which div to apply `overflow: hidden;` on.

**Comment by obenland** [comment]
> I am not able to add new images from Google Photos or Pexels to an existing Gallery block; the option to upload new images comes up first

This one is a little annoying 🙂 
It looks like every click that bubbles out of the External Media components triggers that upload modal because we're rendering everything within that dropzone.

**Comment by obenland** [comment]
> We can merge this PR and address the issues in a follow-up PR though.

@jeherve That would be great. Ajax will be working on getting this ready and a bit more polished over the coming two weeks. Please let me know what we/I can do to make that happen. 

This is the summary of the current status and next steps after your most recent review:
- [x] Rebased after merging #16008 & #16011. 
- [x] It doesn't quite work with the Media & Text block. ✅  Fixed in 0b6a348.
- [x] When editing existing Slideshow blocks, the button seems a bit off. ✅  I _think_ it's fixed with ed6f1e9?
- [ ] Unable to add new images from Google Photos or Pexels to an existing Gallery block. TODO: Prevent click events from bubbling beyond the media selector.
- [ ] Error "invalid authentication credentials". TODO: Handle this wpcom response and return a 403 to re-authenticate. **How can this be reproduced**?
- [ ] After connecting to Google Photos, albums are shown instead of recent photos. TODO: Send the right filters in initial `/list` request.
- [ ] The media insert modal always has scrollbars, which seems a bit weird. TODO: Find a better way to insert images in https://github.com/Automattic/jetpack/issues/15867

> When initially connecting to Google, why is it was offering me ways to disconnect? 

I assume as a courtesy? I've tried to find the code that makes up that modal but was unable to make headway there. It doesn't seem to be in what's covered by opengrok. :(

**Comment by obenland** [comment]
For completeness sake: I created #15988 and #16015, separating out the API endpoint and External Media for only Pexels, in an attempt to make it easier to review/merge. Not sure if it's helpful given how far we've come here already.

**Comment by jeherve** [comment]
> Error "invalid authentication credentials". TODO: Handle this wpcom response and return a 403 to re-authenticate. How can this be reproduced?

I think you'd need to start from an existing, broken Google connection. Mine was created in Calypso more than a year ago.

> For completeness sake: I created #15988 and #16015, separating out the API endpoint and External Media for only Pexels, in an attempt to make it easier to review/merge. Not sure if it's helpful given how far we've come here already.

Let's merge this big one instead, I think that's fine now that you created follow-up items for everything. Thank you!

File: _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+	/**
+	 * Sanitization callback for media parameter.
+	 *
+	 * @param array $param Media parameter.
+	 * @return true|\WP_Error
+	 */
+	public function sanitize_media( $param ) {
+		$param = $this->prepare_media_param( $param );
+
+		return rest_sanitize_value_from_schema( $param, $this->media_schema );
```
Comment (jeherve) [inline _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:160]: Noting that this is currently triggering PHP notices:

```
PHP Notice:  Undefined index: type in /var/www/html/wp-includes/rest-api.php on line 1482
```

Comment (marekhrabe) [reply _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:160]: @jeherve just to be sure, are you fine merging now and addressing the notice separately or should we do it now?

Comment (jeherve) [reply _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-external-media.php:160]: I think we can merge now and address this later 👍

**Comment by jasmussen** [comment]
👏

File: extensions/shared/external-media/editor.scss [hunk 8]
```diff
@@ -0,0 +1,363 @@
+/**
+ * External Media
+ */
+@import '../styles/gutenberg-base-styles.scss';
+
+$grid-size: 8px;
+
+@keyframes loading-fade {
```
Comment (simison) [inline extensions/shared/external-media/editor.scss:8]: @retrofox  Keyframes end up in CSS global scope and should be unique. Recommend prefixing with `jetpack_`.

File: extensions/shared/external-media/editor.scss [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+	}
+}
+
+.jetpack-external-media-filters {
+	display: flex;
+	justify-content: space-between;
+}
+
+// Reset placeholder button margin.
+.components-placeholder__fieldset .components-dropdown .components-button {
```
Comment (simison) [inline extensions/shared/external-media/editor.scss:361]: @retrofox This is not scoped to Jetpack neither and needs fixing.

File: extensions/shared/external-media/editor.scss [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+	}
+
+	.components-modal__content {
+		overflow: auto;
+		padding-bottom: 0;
+	}
+}
+
+@media ( min-width: 600px ) {
+	.components-modal__content {
```
Comment (simison) [inline extensions/shared/external-media/editor.scss:75]: Needs prefixing: https://github.com/Automattic/jetpack/issues/16143#issuecomment-642857130

File: extensions/shared/external-media/editor.scss (hunk 8 above)
Comment (obenland) [reply extensions/shared/external-media/editor.scss:8]: Fixed in #16144.

File: extensions/shared/external-media/editor.scss (hunk 9 above)
Comment (obenland) [reply extensions/shared/external-media/editor.scss:361]: Fixed in #16144.

File: extensions/shared/external-media/editor.scss (hunk 10 above)
Comment (obenland) [reply extensions/shared/external-media/editor.scss:75]: Fixed in #16144

**Pull request merged**